File: annotation-rest/src/main/java/uk/ac/ebi/quickgo/annotation/model/AnnotationRequestBody.java
package uk.ac.ebi.quickgo.annotation.model;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import uk.ac.ebi.quickgo.rest.controller.request.ArrayPattern;

import javax.validation.Valid;
import javax.validation.constraints.Pattern;
import java.util.stream.Stream;

import static uk.ac.ebi.quickgo.annotation.model.AnnotationRequest.DEFAULT_GO_USAGE;
import static uk.ac.ebi.quickgo.annotation.model.AnnotationRequest.USAGE_RELATIONSHIP_PARAM;
import static uk.ac.ebi.quickgo.rest.controller.request.ArrayPattern.Flag.CASE_INSENSITIVE;

@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class AnnotationRequestBody {
  @Valid private GoDescription and;
  @Valid private GoDescription not;

  public static void putDefaultValuesIfAbsent(AnnotationRequestBody requestBody) {
    if (requestBody == null)
      return;
    if (requestBody.getAnd() == null) {
      requestBody.setAnd(new AnnotationRequestBody.GoDescription());
    }
    fillDefaultGoDescriptionIfNotPresent(requestBody.getAnd());

    if (requestBody.getNot() == null) {
      requestBody.setNot(new AnnotationRequestBody.GoDescription());
    }
    fillDefaultGoDescriptionIfNotPresent(requestBody.getNot());

  }

  private static void fillDefaultGoDescriptionIfNotPresent(AnnotationRequestBody.GoDescription goDescription) {
    if (goDescription.getGoTerms() == null) {
      goDescription.setGoTerms(new String[0]);
    }
    if (goDescription.getGoUsage() == null || goDescription.getGoUsage().trim().isEmpty()) {
      goDescription.setGoUsage(DEFAULT_GO_USAGE);
    }
    if (goDescription.getGoUsageRelationships() == null || goDescription.getGoUsageRelationships().length == 0) {
      goDescription.setGoUsageRelationships(AnnotationRequest.DEFAULT_GO_USAGE_RELATIONSHIPS);
    }
  }

  @Data
  @Builder
  @NoArgsConstructor
  @AllArgsConstructor
  public static class GoDescription {
    private String[] goTerms;
    private String[] goUsageRelationships;
    private String goUsage;

    public void setGoUsageRelationships(String goUsageRelationships) {
      String[] arr = goUsageRelationships == null || goUsageRelationships.trim().isEmpty()
        ? new String[]{} : goUsageRelationships.split(",");

      this.goUsageRelationships = Stream.of(arr).map(String::toLowerCase).toArray(String[]::new);
    }

    public void setGoUsage(String goUsage) {
      this.goUsage = goUsage == null ? null : goUsage.toLowerCase();
    }

    @ArrayPattern(regexp = "^GO:[0-9]{7}$", flags = CASE_INSENSITIVE, paramName = "goTerms")
    public String[] getGoTerms() {
      return goTerms;
    }

    @ArrayPattern(regexp = "^is_a|part_of|occurs_in|regulates$", flags = CASE_INSENSITIVE,
      paramName = USAGE_RELATIONSHIP_PARAM)
    public String[] getGoUsageRelationships() {
      return goUsageRelationships;
    }

    @Pattern(regexp = "^slim|descendants|exact$", flags = Pattern.Flag.CASE_INSENSITIVE,
      message = "Invalid goUsage: ${validatedValue}")
    public String getGoUsage() {
      return goUsage;
    }

    public static class GoDescriptionBuilder {
      public GoDescriptionBuilder goUsageRelationships(String... goUsageRelationships) {
        this.goUsageRelationships = Stream.of(goUsageRelationships)
          .map(String::toLowerCase)
          .toArray(String[]::new);
        return this;
      }

      public GoDescriptionBuilder goUsage(String goUsage) {
        this.goUsage = goUsage == null ? null : goUsage.toLowerCase();
        return this;
      }
    }
  }
}

File: annotation-rest/src/main/java/uk/ac/ebi/quickgo/annotation/model/AnnotationRequestBody.kt
package uk.ac.ebi.quickgo.annotation.model


import uk.ac.ebi.quickgo.annotation.model.AnnotationRequest.DEFAULT_GO_USAGE
import uk.ac.ebi.quickgo.annotation.model.AnnotationRequest.USAGE_RELATIONSHIP_PARAM
import uk.ac.ebi.quickgo.rest.controller.request.ArrayPattern
import uk.ac.ebi.quickgo.rest.controller.request.ArrayPattern.Flag.CASE_INSENSITIVE
import javax.validation.Valid
import javax.validation.constraints.Pattern

/**
 * This class is a simple data class with builders having jvm overloads, written using kotlin.
 * Before that we decided to use loombok, but we think for data object/clear code kotlin is more useful
 * Look into git history AnnotationRequestBody.java with loombok annotation deleted (for reference)
 * Might be is future we decide to not use kotlin when java will introduce data objects (release 14 or 15)
 *
 * This class could be lot smaller and simpler but as all test expecting builders, i decided to not make any change at all
 * in other parts of application, that is how we can be sure about the seamless interoperability of kotlin with java
 *
 * For onwards we can simply use data object with jvmoverloads and copy constructor from kotlin to have clean code and easy to test
 */
data class AnnotationRequestBody @JvmOverloads constructor(@field:Valid var and: GoDescription? = null,
                                                           @field:Valid var not: GoDescription? = null) {

    companion object {
        @JvmStatic
        fun builder() = AnnotationRequestBody.Builder()

        @JvmStatic
        fun putDefaultValuesIfAbsent(requestBody: AnnotationRequestBody?) {
            if (requestBody == null)
                return
            if (requestBody.and == null) {
                requestBody.and = AnnotationRequestBody.GoDescription()
            }
            requestBody.and?.let { fillDefaultGoDescriptionIfNotPresent(it) }

            if (requestBody.not == null) {
                requestBody.not = AnnotationRequestBody.GoDescription()
            }
            requestBody.not?.let { fillDefaultGoDescriptionIfNotPresent(it) }
        }

        private fun fillDefaultGoDescriptionIfNotPresent(goDescription: AnnotationRequestBody.GoDescription) {
            val goUsage = goDescription.goUsage
            if (goUsage == null || goUsage.trim().isEmpty()) {
                goDescription.goUsage = DEFAULT_GO_USAGE
            }

            val goUsageRelationships = goDescription.goUsageRelationships
            if (goUsageRelationships == null || goUsageRelationships.isEmpty()) {
                goDescription.setGoUsageRelationships(AnnotationRequest.DEFAULT_GO_USAGE_RELATIONSHIPS)
            }
        }
    }

    data class Builder(
            var and: GoDescription? = null,
            var not: GoDescription? = null) {

        fun and(and: GoDescription) = apply { this.and = and }
        fun not(not: GoDescription) = apply { this.not = not }
        fun build() = AnnotationRequestBody(and, not)
    }


    class GoDescription {

        @JvmOverloads
        constructor (goTerms: Array<String> = emptyArray(),
                     goUsageRelationships: Array<String>? = null,
                     goUsage: String? = null) {
            this.goTerms = goTerms
            this.goUsageRelationships = goUsageRelationships
            this.goUsage = goUsage
        }

        companion object {
            @JvmStatic
            fun builder() = GoDescription.Builder()
        }

        @get:ArrayPattern(regexp = "^GO:[0-9]{7}$", flags = [CASE_INSENSITIVE], paramName = "goTerms")
        var goTerms: Array<String>

        @get:Pattern(regexp = "^descendants|exact$", flags = [Pattern.Flag.CASE_INSENSITIVE], message = "Invalid goUsage: \${validatedValue}")
        var goUsage: String?
            set(goUsage) {
                field = goUsage?.toLowerCase()
            }

        @get:ArrayPattern(regexp = "^is_a|part_of|occurs_in|regulates$", flags = [CASE_INSENSITIVE], paramName = USAGE_RELATIONSHIP_PARAM)
        var goUsageRelationships: Array<String>?
            private set

        fun setGoUsageRelationships(goUsageRelationships: String?) {
            this.goUsageRelationships = goUsageRelationships.orEmpty().split(",")
                    .filter { it.isNotBlank() }
                    .map { it.toLowerCase() }
                    .toTypedArray()
        }

        class Builder {
            private var goTerms: Array<String> = emptyArray()
            private var goUsage: String? = null
            private var goUsageRelationships: Array<String>? = null

            fun goUsageRelationships(vararg goUsageRelationships: String) =
                    apply { this.goUsageRelationships = goUsageRelationships.map { it.toLowerCase() }.toTypedArray(); }

            fun goUsage(goUsage: String?) = apply { this.goUsage = goUsage?.toLowerCase() }
            fun goTerms(goTerms: Array<String>) = apply { this.goTerms = goTerms }
            fun build() = GoDescription(goTerms, goUsageRelationships, goUsage)
        }
    }
}